go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2016, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package planbuilder
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/youtube/vitess/go/vt/sqlparser"
+)
+
```

Comment:
A comment with an overview of the general algorithm would be very helpful. Mention that you're building a binary tree whose nodes satisfy the PlanBuilder interface.